@here/harp-utils/lib/TaskQueue.ts

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright (C) 2017-2020 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * Task that can be added to the [[TaskQueue]]
+ */
+export interface Task {
+    /**
+     * The Function that will executed when the [[Task]] is processed
+     */
+    execute: () => void;
+
+    /**
+     * The group by which similar tasks in the TaskQueue are combined
+     */
+    group: string;
+
+    /**
+     * A function to retrieve the priority of the [[Task]], with 0 being
+     * the highest priority, and the first to be executed
+     */
+    getPrio: () => number;
+
+    /**
+     * An optional function that defines if a [[Task]] is alread expired
+     * and therefore can be removed from the [[TaskQueue]]
+     */
+    isExpired?: () => boolean;
+
+    /**
+     * An optional function that returns an estimated process time,
+     * this is not directly used by the [[TaskQueue]] but can be used
+     * by an Task Scheduler to schedule the processing
+     */
+    estimatedProcessTime?: () => number;
+}
+
+export interface TaskQueueOptions {
+    //TODO: Use the max length
+    //maxLength?: number;
+    /**
+     * Groups to combine specific [[Task]]s in the [[TaskQueue]],
+     * [[Task]]s can only be added to the [[TaskQueue]] if their group is available
+     */
+    groups: string[];
+
+    /**
+     * Optional function to sort the priority, if set, i will override the internal TaskQueue.sort
+     * function.
+     * Caution, the [[TaskQueue]] uses the last element in the Arrays first, so the
+     * highest priorities should be ordered to the end
+     */
+    prioSortFn?: (a: Task, b: Task) => number;
+}
+
+/**
+ * This is a helper interface to be able to index the TaskArrays via a string
+ */
+interface TaskArrays {
+    [key: string]: Task[];
+}
+
+/**
+ * A Pull-TaskQueue sorted by priority and group-able [[Task]]s by [[Task.group]]
+ *
+ * Sample Usage
+ * ```
+ *  const taskQueue = new TaskQueue({
+ *      group: ["group1"]
+ *  })
+ *  taskQueue.add({
+ *     group: "group1",
+ *     execute: () => {
+ *         console.log("task of group1 executed");
+ *     },
+ *     getPrio: () => {
+ *         return 0;
+ *     }
+ *   });
+ *
+ * taskQueue.update();
+ * taskQueue.processNext("group1");
+ *
+ *  ```
+ */
+export class TaskQueue {
+    private m_taskArrays: TaskArrays = {};
+
+    constructor(private m_options: TaskQueueOptions) {
+        this.m_options.groups?.forEach(group => {
+            this.m_taskArrays[group] = [];
+        });
+        this.sort = this.m_options.prioSortFn ?? this.sort;
```

Comment:
`this.sort` is `undefined`

Reply:
this.sort is defined as its a member function   

 private sort(a: Task, b: Task): number {
        // the highest number in the beginning as the last in the array with
        // highest priority which equals 0 will start to be processed
        return b.getPriority() - a.getPriority();
    }

Reply:
ah okay, but to avoid confusion I would only overwrite `this.sort` with `prioSortFn` if it's not defined.

Reply:
Not sure I understand your comment, this.sort will always be defined, as its a member function, if the option.prioSortFn is defined it will overwrite the default this.sort.

Reply:
Yes, but if prioSortFn is undefined you do `this.sort = this.sort`. That is what confused me